flow/raster_cacheable_entry.h

```diff
@@ -0,0 +1,192 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLOW_RASTER_CACHEABLE_ENTRY_H_
+#define FLUTTER_FLOW_RASTER_CACHEABLE_ENTRY_H_
+
+#include <memory>
+#include <optional>
+
+#include "flutter/display_list/display_list_utils.h"
+#include "flutter/flow/embedded_views.h"
+#include "flutter/flow/raster_cache.h"
+#include "include/core/SkCanvas.h"
+#include "include/core/SkMatrix.h"
+#include "include/core/SkPicture.h"
+#include "include/core/SkPoint.h"
+#include "include/core/SkRect.h"
+
+namespace flutter {
+
+class DisplayList;
+class LayerCacheableItem;
+class DisplayListCacheableItem;
+class SkPictureCacheableItem;
+class CacheableContainerLayer;
+
+class CacheableItem {
+ public:
+  enum CacheState {
+    kNone = 0,
+    kCurrent,
+    kChildren,
+  };
+
+  explicit CacheableItem(RasterCacheKeyID key_id,
+                         CacheState cache_state = CacheState::kCurrent,
+                         int item_cache_threshold = 1,
+                         unsigned child_entries = 0)
+      : key_id_(key_id),
+        cache_state_(cache_state),
+        item_cache_threshold_(item_cache_threshold),
+        child_entries_(child_entries) {}
+
+  virtual void PrerollSetup(PrerollContext* context,
+                            const SkMatrix& matrix) = 0;
+
+  virtual void PrerollFinalize(PrerollContext* context,
+                               const SkMatrix& matrix) = 0;
+
+  virtual bool Draw(const PaintContext& context,
+                    const SkPaint* paint) const = 0;
+
+  void Touch(const RasterCache*);
+
+  virtual std::optional<RasterCacheKeyID> GetId() const { return key_id_; }
+
+  virtual LayerCacheableItem* asLayerCacheableItem() { return nullptr; }
+
+  virtual DisplayListCacheableItem* asDisplayCacheableItem() { return nullptr; }
+
+  virtual SkPictureCacheableItem* asSkPictureCacheableItem() { return nullptr; }
+
+  virtual bool TryToPrepareRasterCache(const PaintContext& context) const = 0;
+
+  unsigned chld_entries() const { return child_entries_; }
+
+  void set_matrix(const SkMatrix& matrix) { matrix_ = matrix; }
+
+  void set_child_entries(unsigned child_entries) {
+    child_entries_ = child_entries;
+  }
+
+  bool need_cached() const { return cache_state_ != CacheState::kNone; }
+
+  virtual ~CacheableItem() = default;
+
+ protected:
+  // The id for cache the layer self.
+  RasterCacheKeyID key_id_;
+  CacheState cache_state_ = CacheState::kCurrent;
+  int item_cache_threshold_ = 1;
+  int num_cache_attempts_ = 0;
+  mutable SkMatrix matrix_;
+  unsigned child_entries_;
+};
+
+class LayerCacheableItem : public CacheableItem {
```

Comment:
I'd move these types into different files. Ganging up like this makes it harder to review code for parallel implementations of an interface. In general, unless it is a small helper class for the main class in a file, classes should get their own files for isolation and clarity.

Reply:
I was also debating whether these files should live here or in the flow/layers directory where they might end up nearer to the code that uses them.

Layers use CacheableItems which use the RasterCache.

The RasterCache is (becoming) a neutral Keys to Images cache with just a few items in the Entry field for bookkeeping.